frontend/src/metabase/actions/containers/ActionCreatorModal/ActionCreatorModal.unit.spec.tsx

```diff
@@ -0,0 +1,110 @@
+import React from "react";
+import { Route } from "react-router";
+import nock from "nock";
+
+import {
+  renderWithProviders,
+  screen,
+  waitForElementToBeRemoved,
+} from "__support__/ui";
+import {
+  setupActionsEndpoints,
+  setupCardsEndpoints,
+} from "__support__/server-mocks";
+
+import * as Urls from "metabase/lib/urls";
+
+import type { Card, WritebackAction } from "metabase-types/api";
+import {
+  createMockCard,
+  createMockQueryAction,
+} from "metabase-types/api/mocks";
+
+import ActionCreatorModal from "./ActionCreatorModal";
+
+jest.mock(
+  "metabase/actions/containers/ActionCreator",
+  () =>
+    function MockActionCreator() {
+      return <div data-testid="mock-action-creator" />;
+    },
+);
+
+const MODEL = createMockCard({ id: 1, dataset: true });
+const ACTION = createMockQueryAction({ model_id: MODEL.id });
+const ACTION_NOT_FOUND_ID = 999;
+
+type SetupOpts = {
+  initialRoute: string;
+  action?: WritebackAction | null;
+  model?: Card;
+};
+
+async function setup({
+  initialRoute,
+  model = MODEL,
+  action = ACTION,
+}: SetupOpts) {
+  const scope = nock(location.origin);
+
+  setupCardsEndpoints(scope, [model]);
+
+  if (action) {
+    setupActionsEndpoints(scope, model.id, [action]);
+  } else {
+    scope.get(`/api/action/${ACTION_NOT_FOUND_ID}`).reply(404);
+  }
+
+  const { history } = renderWithProviders(
+    <>
+      <Route
+        path="/model/:slug/detail/actions/:actionId"
+        component={ActionCreatorModal}
+      />
+      <Route
+        path="/model/:slug/detail/actions"
+        component={() => <div data-testid="mock-model-detail" />}
+      />
+    </>,
+    {
+      withRouter: true,
+      initialRoute,
+    },
+  );
+
+  await waitForElementToBeRemoved(() => screen.queryAllByText(/Loading/i));
+
+  return { history };
+}
+
+describe("actions > containers > ActionCreatorModal", () => {
+  it("renders correctly", async () => {
+    await setup({ initialRoute: Urls.action(MODEL, ACTION.id) });
+    expect(
+      await screen.findByTestId("mock-action-creator"),
+    ).toBeInTheDocument();
+  });
+
+  it("redirects back to the model detail page if the action is not found", async () => {
+    const initialRoute = Urls.action(MODEL, ACTION_NOT_FOUND_ID);
+    const { history } = await setup({ initialRoute, action: null });
+
+    expect(await screen.findByTestId("mock-model-detail")).toBeInTheDocument();
+    expect(screen.queryByTestId("mock-action-creator")).not.toBeInTheDocument();
+    expect(history?.getCurrentLocation().pathname).toBe(
+      Urls.modelDetail(MODEL, "actions"),
```

Comment:
sometimes I'm on the fence about using the implementation in the test assertions.
1. It's hard to visualize what the value is, I still think we should try to stick to black-box testing more, like asserting the expected shape e.g. raw string or object, etc.
2. We should focus on testing the behavior more. I know these are unit tests, but our unit tests are more like integration tests when we could actually render the whole application with certain parts being mocked. We shouldn't be using the actual code as an assertion too much IMO. Well if your code uses the exact function it's gonna pass, isn't it? But it isn't guaranteed that this resolves to the actually expected path.

What do you think?

Reply:
Yeah, I think you're right and we should better hardcode the URL here 👍